packages/primitives/src/components/Combobox/Combobox.tsx

```diff
@@ -0,0 +1,764 @@
+  /**
+   * Event handler called when the escape key is down.
+   * Can be prevented.
+   */
+  onEscapeKeyDown?: DismissableLayerProps['onEscapeKeyDown'];
+  /**
+   * Event handler called when the a `pointerdown` event happens outside of the `DismissableLayer`.
+   * Can be prevented.
+   */
+  onPointerDownOutside?: DismissableLayerProps['onPointerDownOutside'];
+
+  position?: 'item-aligned' | 'popper';
+}
+
+/**
+ * @internal This is the implementation of the `ComboboxContent` component. But should not be used on it's own
+ * for accessibility reasons. Use `ComboboxContent` instead.
+ */
+const ComboboxContentImpl = React.forwardRef<ComboboxContentImplElement, ComboboxContentImplProps>(
+  (props, forwardedRef) => {
+    const { position = 'item-aligned', onEscapeKeyDown, onPointerDownOutside, ...contentProps } = props;
+    const context = useComboboxContext(CONTENT_NAME);
+    const composedRefs = useComposedRefs(forwardedRef, (node) => context.onContentChange(node));
+
+    // prevent selecting items on `pointerup` in some cases after opening from `pointerdown`
+    // and close on `pointerup` outside.
+    const { onOpenChange } = context;
+
+    const getItems = useCollection(undefined);
+
+    React.useEffect(() => {
+      const close = () => onOpenChange(false);
+      window.addEventListener('blur', close);
+      window.addEventListener('resize', close);
+      return () => {
+        window.removeEventListener('blur', close);
+        window.removeEventListener('resize', close);
+      };
+    }, [onOpenChange]);
+
+    return (
+      <RemoveScroll allowPinchZoom>
+        <DismissableLayer
+          asChild
+          onEscapeKeyDown={onEscapeKeyDown}
+          onPointerDownOutside={onPointerDownOutside}
+          // When focus is trapped, a focusout event may still happen.
+          // We make sure we don't trigger our `onDismiss` in such case.
+          onFocusOutside={(event) => event.preventDefault()}
+          onDismiss={() => {
+            context.onOpenChange(false);
+            context.trigger?.focus({ preventScroll: true });
+          }}
+        >
+          <ComboboxPopperPosition
+            role="listbox"
+            id={context.contentId}
+            data-state={context.open ? 'open' : 'closed'}
+            onContextMenu={(event) => event.preventDefault()}
+            {...contentProps}
+            ref={composedRefs}
+            style={{
+              // flex layout so we can place the scroll buttons properly
+              display: 'flex',
+              flexDirection: 'column',
+              // reset the outline by default as the content MAY get focused
+              outline: 'none',
+              ...contentProps.style,
+            }}
+            onKeyDown={composeEventHandlers(contentProps.onKeyDown, (event) => {
+              // Combobox should not be navigated using tab key so we prevent it
+              if (event.key === 'Tab') event.preventDefault();
```

Comment:
Since this is handled for all cases at the end I don't think you need to handle this separately?

Reply:
It's not handled for all cases, just `['ArrowUp', 'ArrowDown', 'Home', 'End']`

Reply:
Ah missed the closing bracket. Mh - what do you think of grouping these together?

```
if (['ArrowUp', 'ArrowDown', 'Home', 'End', 'Tab'].includes(event.key)) {
  event.preventDefault();
}
```

at the end?

Reply:
I don't really mind but I found it easier to understand:

* this key does that
* these keys do this

instead of:
* these keys do this
* all these keys prevent default though

What makes you prefer the latter?

Reply:
I personally find the latter case you've described easier to understand, but since is just a coding-style thing and your solution is in-line with all the other places let's keep it 👍🏼